File: EgoCapture/Services/Packaging/ThumbnailGenerator.swift
import AVFoundation
import UIKit

/// Extracts the first frame of a session's `video_{code}.mp4` and caches it as
/// `thumbnail_{code}.jpg` inside the session directory. The Sessions screen
/// reads the cached file synchronously — it falls back to a gradient
/// placeholder while the async generation completes (or when the video file
/// is missing / still uploading).
///
/// The generator always writes 128×128 JPEGs (~10–20 KB) via aspect-fill
/// cropping, which keeps the per-session disk cost negligible compared to
/// the original video.
enum ThumbnailGenerator {

    /// Canonical location of the cached JPEG for a given session directory.
    static func cacheURL(in sessionDir: URL) -> URL {
        SessionFiles.url("thumbnail", "jpg", in: sessionDir)
    }

    /// Return the cached thumbnail synchronously if it's already on disk.
    /// Callers use this for the first render, then fire `generateIfNeeded`
    /// to backfill when it returns `nil`.
    static func cachedImage(in sessionDir: URL) -> UIImage? {
        let url = cacheURL(in: sessionDir)
        guard FileManager.default.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url),
              let image = UIImage(data: data) else { return nil }
        return image
    }

    /// Generate the thumbnail for `sessionDir` if it isn't cached yet. Safe to
    /// call repeatedly — subsequent calls short-circuit on the disk cache.
    /// Returns `nil` when the source video is missing or AVFoundation fails
    /// to decode the first sample.
    @discardableResult
    static func generateIfNeeded(in sessionDir: URL) async -> UIImage? {
        if let cached = cachedImage(in: sessionDir) { return cached }

        guard let videoURL = SessionFiles.resolveExisting("video", "mp4", in: sessionDir) else {
            return nil
        }

        return await extract(from: videoURL, cachingAt: cacheURL(in: sessionDir))
    }

    // MARK: - Private

    private static func extract(from videoURL: URL, cachingAt cacheURL: URL) async -> UIImage? {
        let asset = AVURLAsset(
            url: videoURL,
            options: [AVURLAssetPreferPreciseDurationAndTimingKey: true]
        )

        let generator = AVAssetImageGenerator(asset: asset)
        generator.appliesPreferredTrackTransform = true
        // Give AVFoundation room to pick the closest keyframe without
        // re-decoding hundreds of frames — we only need a ballpark "first frame".
        generator.requestedTimeToleranceBefore = CMTime(seconds: 0.25, preferredTimescale: 600)
        generator.requestedTimeToleranceAfter  = CMTime(seconds: 0.25, preferredTimescale: 600)
        generator.maximumSize = CGSize(width: 256, height: 256)

        // 100 ms into the clip avoids the occasional green / black first frame
        // some encoders emit when the IDR is being assembled.
        let time = CMTime(seconds: 0.1, preferredTimescale: 600)

        let cgImage: CGImage?
        if #available(iOS 16.0, *) {
            cgImage = try? await generator.image(at: time).image
        } else {
            cgImage = try? generator.copyCGImage(at: time, actualTime: nil)
        }

        guard let cg = cgImage else { return nil }

        let rendered = aspectFillResize(
            UIImage(cgImage: cg),
            to: CGSize(width: 128, height: 128)
        )

        if let data = rendered.jpegData(compressionQuality: 0.8) {
            try? data.write(to: cacheURL, options: .atomic)
        }

        return rendered
    }

    /// Center-crop + scale to fit `size` exactly (aspect-fill). Avoids the
    /// letter-boxing that `scaledToFit` would introduce on 16:9 / 4:3 frames.
    private static func aspectFillResize(_ image: UIImage, to size: CGSize) -> UIImage {
        let srcW = image.size.width
        let srcH = image.size.height
        guard srcW > 0, srcH > 0 else { return image }

        let scale = max(size.width / srcW, size.height / srcH)
        let targetW = srcW * scale
        let targetH = srcH * scale
        let origin = CGPoint(
            x: (size.width - targetW) / 2,
            y: (size.height - targetH) / 2
        )

        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        let renderer = UIGraphicsImageRenderer(size: size, format: format)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: origin, size: CGSize(width: targetW, height: targetH)))
        }
    }
}

File: android/app/src/main/java/com/kgeneye/eye/session/ThumbnailGenerator.kt
package com.kgeneye.eye.session

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.media.MediaMetadataRetriever
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import kotlin.math.max

/**
 * Extracts the first frame of a session's `video_<code>.mp4` and caches it
 * as `thumbnail_<code>.jpg` (128×128 aspect-fill JPEG, q=0.8) inside the
 * session directory. Mirrors `ThumbnailGenerator` on iOS so the Sessions
 * screen shows a real frame as soon as the recording finishes.
 *
 * Safe to call repeatedly — subsequent calls short-circuit on the disk
 * cache. Returns `null` if the source video is missing or
 * `MediaMetadataRetriever` fails to decode the early sample.
 */
object ThumbnailGenerator {

    private const val THUMB_SIZE = 128
    private const val JPEG_QUALITY = 80

    fun cacheFile(sessionDir: File): File =
        SessionFiles.file("thumbnail", "jpg", sessionDir)

    fun cachedBitmap(sessionDir: File): Bitmap? {
        val f = cacheFile(sessionDir)
        if (!f.exists()) return null
        return try { BitmapFactory.decodeFile(f.absolutePath) } catch (_: Throwable) { null }
    }

    /**
     * Generate the thumbnail if it isn't cached yet. Looks up the canonical
     * `video_<code>.mp4` in [sessionDir] and writes
     * `thumbnail_<code>.jpg` next to it.
     */
    fun generateIfNeeded(sessionDir: File): Bitmap? {
        val cache = cacheFile(sessionDir)
        if (cache.exists()) {
            return try { BitmapFactory.decodeFile(cache.absolutePath) } catch (_: Throwable) { null }
        }
        val video = SessionFiles.file("video", "mp4", sessionDir)
        if (!video.exists()) return null

        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(video.absolutePath)
            // 100 ms in to avoid the occasional black/green opening frame.
            val frame = retriever.getFrameAtTime(
                100_000L,
                MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
            ) ?: return null
            val resized = aspectFillResize(frame, THUMB_SIZE, THUMB_SIZE)
            FileOutputStream(cache).use { out ->
                resized.compress(Bitmap.CompressFormat.JPEG, JPEG_QUALITY, out)
            }
            if (resized !== frame) frame.recycle()
            resized
        } catch (t: Throwable) {
            Log.w(TAG, "thumbnail generation failed: ${t.message}")
            null
        } finally {
            try { retriever.release() } catch (_: Throwable) {}
        }
    }

    /**
     * Center-crop + scale so the source covers the entire `target × target`
     * canvas (aspect-fill). Avoids the letter-boxing `Bitmap.createScaledBitmap`
     * would introduce for 16:9 / 4:3 frames.
     */
    private fun aspectFillResize(src: Bitmap, targetW: Int, targetH: Int): Bitmap {
        val srcW = src.width
        val srcH = src.height
        if (srcW <= 0 || srcH <= 0) return src
        val scale = max(targetW.toDouble() / srcW, targetH.toDouble() / srcH)
        val drawW = (srcW * scale).toInt().coerceAtLeast(1)
        val drawH = (srcH * scale).toInt().coerceAtLeast(1)
        val dx = (targetW - drawW) / 2
        val dy = (targetH - drawH) / 2

        val out = Bitmap.createBitmap(targetW, targetH, Bitmap.Config.ARGB_8888)
        Canvas(out).apply {
            drawColor(0xFF000000.toInt())
            drawBitmap(
                src,
                Rect(0, 0, srcW, srcH),
                Rect(dx, dy, dx + drawW, dy + drawH),
                Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG),
            )
        }
        return out
    }

    private const val TAG = "ThumbnailGenerator"
}
